crates/adapters/src/format/parquet/mod.rs

```diff
@@ -0,0 +1,367 @@
+            "time" => DataType::Time64(TimeUnit::Nanosecond),
+            "timestamp" => DataType::Timestamp(TimeUnit::Millisecond, None),
+            "date" => DataType::Date32,
+            "array" => unimplemented!("handle array types"),
+            s => unimplemented!("Encountered unknown type {}", s),
+        }
+    }
+
+    let fields = relation
+        .fields
+        .iter()
+        .map(|f| {
+            Field::new(
+                &f.name,
+                columntype_to_datatype(&f.columntype),
+                f.columntype.nullable,
+            )
+        })
+        .collect::<Vec<Field>>();
+
+    // TODO: error handling, not clear how this can fail
+    SerdeArrowSchema::from_arrow_fields(&fields).expect("Failed to create arrow schema")
+}
+
+struct ParquetEncoder {
+    /// Input handle to push serialized data to.
+    output_consumer: Box<dyn OutputConsumer>,
+    _relation: Relation,
+    parquet_schema: SerdeArrowSchema,
+    config: ParquetEncoderConfig,
+    buffer: Vec<u8>,
+    max_buffer_size: usize,
+}
+
+impl ParquetEncoder {
+    fn new(
+        output_consumer: Box<dyn OutputConsumer>,
+        config: ParquetEncoderConfig,
+        _relation: Relation,
+    ) -> Self {
+        let max_buffer_size = output_consumer.max_buffer_size_bytes();
+        Self {
+            output_consumer,
+            config,
+            parquet_schema: relation_to_parquet_schema(&_relation),
+            _relation,
+            buffer: Vec::new(),
+            max_buffer_size,
+        }
+    }
+}
+
+impl Encoder for ParquetEncoder {
+    fn consumer(&mut self) -> &mut dyn OutputConsumer {
+        self.output_consumer.as_mut()
+    }
+
+    fn encode(&mut self, batches: &[Arc<dyn SerBatch>]) -> AnyResult<()> {
+        let mut buffer = take(&mut self.buffer);
+        let props = WriterProperties::builder().build();
+        let schema = Arc::new(Schema::new(self.parquet_schema.to_arrow_fields()?));
+        let fields = self.parquet_schema.to_arrow_fields()?;
+        let mut builder = ArrowBuilder::new(&fields)?;
+
+        let mut num_records = 0;
+        for batch in batches.iter() {
+            let mut cursor = CursorWithPolarity::new(
+                batch.cursor(RecordFormat::Parquet(self.parquet_schema.clone()))?,
+            );
+            while cursor.key_valid() {
+                if !cursor.val_valid() {
+                    cursor.step_key();
+                    continue;
+                }
+                let mut w = cursor.weight();
+                if !(-MAX_DUPLICATES..=MAX_DUPLICATES).contains(&w) {
+                    bail!("Unable to output record with very large weight {w}. Consider adjusting your SQL queries to avoid duplicate output records, e.g., using 'SELECT DISTINCT'.");
+                }
+
+                while w != 0 {
+                    let prev_len = buffer.len();
+                    cursor.serialize_arrow_key(&mut builder)?;
```

Comment:
We're ignoring `w` here, which means that inserts and deletes are indistinguishable. A slightly safer choice is to skip deletes.